fix(ngMessages): ensure that multi-level transclusion works with ngMessagesInclude

matsko opened the pull request:
ngRepeat and any other directives that alter the DOM structure using
transclusion may cause ngMessagesInclude to behave in an unpredictable
manner. This fix ensures that the element containing the ngMessagesInclude
directive will stay in the DOM to avoid these issues.

Closes #11196

caitp commented on src/ngMessages/messages.js:
```diff
@@ -496,31 +496,16 @@ angular.module('ngMessages', [])
+       link: function($scope, element, attrs) {
+         $templateRequest(attrs.ngMessagesInclude || attrs.src).then(function(html) {
+           element.html(html);
+           $compile(element.contents())($scope);
```
I'm not sure this is the best approach. Why not just make use of the clone attach feature of the link function?

```
$templateRequest(attrs.ngMessagesInclude || attrs.src).then(function(html) {
  $compile(html)($scope, function(dom) {
    // attach to DOM using your favourite method --- finding the parent ngMessages directive will no longer fail.
  });
```

matsko replied:
What's the benefit to having this code be inserted in a different way? There's no animation stuff here since that's managed in the `<ng-message>` directive level which transcludes on its own anyway.

caitp replied:
the benefit is not adding a breaking change --- keeping the DOM looking the way it did before

matsko replied:
That works amazing! Thanks! You just saved a very inconvenient breaking change.

caitp commented on src/ngMessages/messages.js:
```diff
@@ -496,31 +496,16 @@ angular.module('ngMessages', [])
-           $templateRequest(attrs.ngMessagesInclude || attrs.src).then(function(html) {
-             var elements = jqLite('<div></div>').html(html).contents();
-             var cursor = comment;
-             forEach(elements, function(elm) {
```
+1 for removing all this though, it's more complicated than is really needed

caitp commented on src/ngMessages/messages.js:
```diff
@@ -496,31 +496,16 @@ angular.module('ngMessages', [])
-       require: '^^ngMessages',
-       compile: function(element, attrs) {
-         var comment = jqLite($document[0].createComment(' ngMessagesInclude: '));
-         element.after(comment);
```
making this directive an element transclusion directive makes it unnecessary to add this manually

matsko replied:
If we set terminal to true then will this run last if the priority is left unset?

caitp replied:
directives are evaluated in the deterministic order, but they'll stop when they reach the terminal one. similarly, child nodes of an element with a terminal directive won't be compiled (but there are no children, except the ones you load from the server and compile on your own.)

matsko replied:
Yes I know how this works, but please answer my question if it will be handled last if a priority is not assigned in the directive definition.

caitp replied:
yes, it will. the default priority is 0

matsko replied:
The PR has been updated with a new commit. I don't think we should set this to terminal since it is already working.

matsko committed:
fix(ngMessages): ensure that multi-level transclusion works with ngMessagesInclude

ngRepeat and any other directives that alter the DOM structure using
transclusion may cause ngMessagesInclude to behave in an unpredictable
manner. This fix ensures that the element containing the ngMessagesInclude
directive will stay in the DOM to avoid these issues.

Closes #11196

caitp commented:
LGTM, but I would suggest adding some comments explaining the flow and assertions in that large unit test.

caitp commented on src/ngMessages/messages.js:
```diff
@@ -500,27 +500,13 @@ angular.module('ngMessages', [])
-               cursor = elm;
-             });
-             $compile(elements)($scope);
+       require: '^^ngMessages', // we only require this for validation sake
```
actually, hang on. Previously, you were basically replacing the element with a comment placeholder. I think we should keep doing this, there's no good reason to keep it in the DOM (is there?)

matsko replied:
All fixed up now. And I added some comments to the unit test.

caitp commented:
thanks, lgtm.

matsko commented:
Merged as https://github.com/angular/angular.js/commit/d7ec5f392e1550658ddf271a30627b1749eccb69

matsko commented:
Thanks @caitp
